lib/stm32/f3/opamp.c

```diff
@@ -0,0 +1,156 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**@{*/
+
+#include <libopencm3/stm32/opamp.h>
+
+void opamp_enable(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_EN;
+}
+
+void opamp_disable(uint32_t base)
+{
+	OPAMP_CSR(base) &= ~OPAMP_CSR_EN;
+}
+
+void opamp_lock(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_LOCK;
+}
+
+
+void opamp_set_calsel(uint32_t base, uint32_t calsel)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_CALSEL_MASK << OPAMP_CSR_CALSEL_SHIFT);
+	OPAMP_CSR(base) |= calsel << OPAMP_CSR_CALSEL_SHIFT;
+}
+
+bool opamp_read_outcal(uint32_t base)
+{
+	return (OPAMP_CSR(base) >> OPAMP_CSR_OUTCAL_SHIFT)
+		& OPAMP_CSR_OUTCAL_MASK;
+}
+
+void opamp_tcm_enable(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_TCM_EN;
+}
+
+void opamp_tcm_disable(uint32_t base)
+{
+	OPAMP_CSR(base) &= ~OPAMP_CSR_TCM_EN;
+}
+
+void opamp_force_vp_enable(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_FORCE_VP;
+}
+
+void opamp_force_vp_disable(uint32_t base)
+{
+	OPAMP_CSR(base) &= ~OPAMP_CSR_FORCE_VP;
+}
+
+void opamp_cal_enable(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_CALON;
+}
+
+void opamp_cal_disable(uint32_t base)
+{
+	OPAMP_CSR(base) &= ~OPAMP_CSR_CALON;
+}
+
+
+void opamp_trimoffsetn_set(uint32_t base, uint32_t trim)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_TRIMOFFSETN_MASK
+		<< OPAMP_CSR_TRIMOFFSETN_SHIFT);
+	OPAMP_CSR(base) |= trim << OPAMP_CSR_TRIMOFFSETN_SHIFT;
+}
+
+void opamp_trimoffsetp_set(uint32_t base, uint32_t trim)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_TRIMOFFSETP_MASK
+		<< OPAMP_CSR_TRIMOFFSETP_SHIFT);
+	OPAMP_CSR(base) |= trim << OPAMP_CSR_TRIMOFFSETP_SHIFT;
+}
+
+void opamp_user_trim_enable(uint32_t base)
+{
+	OPAMP_CSR(base) |= OPAMP_CSR_USER_TRIM;
+}
+
+void opamp_user_trim_disable(uint32_t base)
+{
+	OPAMP_CSR(base) &= ~OPAMP_CSR_USER_TRIM;
+}
+
+
+void opamp_pga_gain_select(uint32_t base, uint32_t gain)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_PGA_GAIN_MASK
+		<< OPAMP_CSR_PGA_GAIN_SHIFT);
+	OPAMP_CSR(base) |= gain << OPAMP_CSR_PGA_GAIN_SHIFT;
+}
+
+
+void opamp_vp_select(uint32_t base, uint32_t vp)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_VP_SEL_MASK
+		<< OPAMP_CSR_VP_SEL_SHIFT);
+	OPAMP_CSR(base) |= vp << OPAMP_CSR_VP_SEL_SHIFT;
+}
+
+void opamp_vm_select(uint32_t base, uint32_t vm)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_VM_SEL_MASK
+		<< OPAMP_CSR_VM_SEL_SHIFT);
+	OPAMP_CSR(base) |= vm << OPAMP_CSR_VM_SEL_SHIFT;
+}
+
+
+void opamp_vps_select(uint32_t base, uint32_t vps)
+{
+	OPAMP_CSR(base) &= ~(OPAMP_CSR_VPS_SEL_MASK
+		<< OPAMP_CSR_VPS_SEL_SHIFT);
+	OPAMP_CSR(base) |= vps << OPAMP_CSR_VPS_SEL_SHIFT;
+}
+
+void opamp_vms_select(uint32_t base, uint32_t vms)
+{
```

Comment:
why _select on all of these, with _set for others?

Reply:
I used select wherever the input was expected to be an enum, and set where it's a value. Saw that kind of thing elsewhere and assumed, but am happy to change them all to set if that makes it more consistent?